### PR TITLE
Benchmarking instrumentation

### LINKED ISSUE #485: -DHERA_BENCHMARKING=ON
Can we add flag `-DHERA_BENCHMARKING=ON` when compiling Hera, and include the following code in each engine's interface to print benchmarking data? @chfast 

```
#if HERA_BENCHMARKING
#include <fstream>      // std::fstream
#include <time.h>       // clock_t, clock
#endif

...

#if HERA_BENCHMARKING
clock_t t = clock();
#endif

...parse/validate/compile...

#if HERA_BENCHMARKING
t = clock() - t;
std::fstream fs;
fs.open ("runtime_data_wavm_compile.csv", std::fstream::out | std::fstream::app);
fs << ", " << ((float)t)/CLOCKS_PER_SEC;
fs.close();
#endif

...

#if HERA_BENCHMARKING
clock_t t2 = clock();
#endif

...invoke main...

#if HERA_BENCHMARKING
t2 = clock() - t2;
std::fstream fs2;
fs2.open ("runtime_data_wavm_invoke.csv", std::fstream::out | std::fstream::app);
fs2 << ", " << ((float)t2)/CLOCKS_PER_SEC;
fs2.close();
#endif
```

### PR DESCRIPTION
**Pull request by chfast**
Closes #485.

### CODE REVIEW

File: src/eei.h [hunk 1]
```diff
@@ -47,6 +50,42 @@ class WasmEngine {
...
+
+  void executionFinished() noexcept
+  {
+    if (benchmarkingEnabled)
+    {
+      constexpr auto to_us = [](clock::duration d) {
+        return std::chrono::duration_cast<std::chrono::microseconds>(d).count();
+      };
+
+      auto executionDuration = clock::now() - executionStartTime;
```
Comment (poemm) [inline src/eei.h]: `executionDuration` is not interesting to me other than to compute invocation time as `executionDuration - instantiationDuration`. But `instantiationDuration` is interesting.

File: src/hera.cpp [hunk 2]
```diff
@@ -398,6 +398,14 @@ evmc_set_option_result hera_set_option(
     return EVMC_SET_OPTION_SUCCESS;
   }
 
+  if (strcmp(name, "benchmark") == 0) {
```
Comment (poemm) [inline src/hera.cpp]: So we would run testeth with command line argument `benchmarking=true`?

**Comment by poemm** [comment]
Thank you for this. Looks good. I see no problems. @axic do you insist on preprocessor flags?

**Comment by axic** [comment]
> @axic do you insist on preprocessor flags?

Does it matter?

**Comment by poemm** [comment]
> Does it matter?

Hmm. Depends on who you ask. This PR is OK for me. But I respect your design goals, so if you want preprocessor flags, then we will use preprocessor flags. I am blocked on this so I hope that you give guidance soon.

File: src/hera.cpp (hunk 2 above)
Comment (chfast) [reply src/hera.cpp]: Yes, by `--evmc benchmark=true`.

**Comment by poemm** [comment]
Gah, benchmarking already has scripts to parse the previous stderr statements, my fault for not telling you. Thank you for improving the stderr statements, it may be helpful to include the words "instantiation" and "execution" to make it understandable, but after that please leave the stderr statements so that I only have to write my parsing script once. Also, the file may be useful, but please make it print the same thing as stderr.

BTW this design overwrites `instantiationStartTime` when a contract calls another contract. Which is OK for benchmarking since benchmarks are one contract calling the contract to be benchmarked.

File: src/eei.cpp [hunk 3]
```diff
@@ -45,6 +46,25 @@ bool exceedsUint128(evmc_uint256be const& value) noexcept
 }
 }  // namespace
 
+bool WasmEngine::benchmarkingEnabled = false;
```
Comment (axic) [inline src/eei.cpp]: I think this should go into the header.

File: src/eei.cpp [hunk 4]
```diff
@@ -45,6 +46,25 @@ bool exceedsUint128(evmc_uint256be const& value) noexcept
...
+  };
+
+  const auto now = clock::now();
+  const auto instantiationDuration = executionStartTime - instantiationStartTime;
+  const auto executionDuration = now - executionStartTime;
+
+  std::cerr << "Time [us]: " << to_us(instantiationDuration + executionDuration) << " = "
+            << to_us(instantiationDuration) << " + " << to_us(executionDuration) << "\n";
+
+  std::ofstream f{"hera_benchmarks.csv", std::ios::out | std::ios::app};
```
Comment (axic) [inline src/eei.cpp]: If we're creating files I think that should be documented in the readme.

**Comment by axic** [comment]
> BTW this design overwrites `instantiationStartTime` when a contract calls another contract. Which is OK for benchmarking since benchmarks are one contract calling the contract to be benchmarked.

I don't think so. Each execution is a new instance of the engine. The only problem is in the output the children come first.

File: src/eei.cpp (hunk 3 above)
Comment (chfast) [reply src/eei.cpp]: That would define the variable multiple times.

Comment (axic) [reply src/eei.cpp]: Why?

Comment (chfast) [reply src/eei.cpp]: Because the header is included in multiple compilation units. You will not be able to link it in the end.

File: src/eei.h [hunk 5]
```diff
@@ -47,6 +50,35 @@ class WasmEngine {
...
+  {
+    if (benchmarkingEnabled)
+      collectBenchmarkingData();
+  }
+
+private:
+  void collectBenchmarkingData();
+
+  using clock = std::chrono::high_resolution_clock;
+  static bool benchmarkingEnabled;
```
Comment (axic) [inline src/eei.h]: I'm suggesting to set this here to `false`.

File: src/eei.cpp (hunk 3 above)
Comment (axic) [reply src/eei.cpp]: Maybe there's some confusion here. I do not suggest to move this statement as is, but: https://github.com/ewasm/hera/pull/486/files#r256387031

File: src/eei.h (hunk 5 above)
Comment (poemm) [reply src/eei.h]: Static member variables must be initialized outside of class.

File: src/eei.cpp [hunk 6]
```diff
@@ -45,6 +46,25 @@ bool exceedsUint128(evmc_uint256be const& value) noexcept
...
+  constexpr auto to_us_str = [](clock::duration d) {
+    return std::to_string(std::chrono::duration_cast<std::chrono::microseconds>(d).count());
+  };
+
+  const auto now = clock::now();
+  const auto instantiationDuration = executionStartTime - instantiationStartTime;
+  const auto executionDuration = now - executionStartTime;
+
+  const auto log = "Time [us]: " + to_us_str(instantiationDuration + executionDuration) + " = " +
+                   to_us_str(instantiationDuration) + " + " + to_us_str(executionDuration) + "\n";
```
Comment (poemm) [inline src/eei.cpp]: Should we print words like "instantiation" and "invocation" here?

**Comment by poemm** [comment]
> I don't think so. Each execution is a new instance of the engine. The only problem is in the output the children come first.

You are correct.

File: src/eei.h (hunk 5 above)
Comment (chfast) [reply src/eei.h]:
```
/home/chfast/Projects/ewasm/hera/src/eei.h:79:15: error: ISO C++ forbids in-class initialization of non-const static member ‘hera::WasmEngine::benchmarkingEnabled’
   static bool benchmarkingEnabled = false;
               ^~~~~~~~~~~~~~~~~~~
```

File: src/eei.cpp [hunk 7]
```diff
@@ -45,6 +46,25 @@ bool exceedsUint128(evmc_uint256be const& value) noexcept
...
+  std::cerr << log;
+  std::ofstream{"hera_benchmarks.log", std::ios::out | std::ios::app} << log;
```
Comment (axic) [inline src/eei.cpp]: To solve the problem of calls showing the times for children first, perhaps you could pass `msg` to `collectBenchmarkingData` and include some identifiers in the log file, such as `msg.depth`, `msg.sender`, `msg.destination`.

Comment (poemm) [reply src/eei.cpp]: This is a good idea. But this is unnecessary for all current benchmarks which either (i) execute one function, or (ii) a bencher function calls the function to be benchmarked, possibly multiple times.

**Comment by chfast** [comment]
How about this output format:
```
Time [us]: NNN (instantiation: XXX, execution: YYY)
```

**Comment by poemm** [comment]
Tiny things: it would be easier to parse if `XXX` did not touch `,` and `YYY` did not touch `)`.

**Comment by axic** [comment]
> it would be easier to parse if `XXX` did not touch `,` and `YYY` did not touch `)`.

Why? Actually that should make it easier to parse with regexp.

**Comment by axic** [comment]
Needs a rebase.

**Comment by poemm** [comment]
> Why? Actually that should make it easier to parse with regexp.

Nevermind, I can parse it easily like it is now.

**Comment by chfast** [comment]
Ready?

File: src/eei.cpp (hunk 6 above)
Comment (poemm) [reply src/eei.cpp]: For me this is ok, but for others it may be polite to explain the numbers.

**Comment by chfast** [comment]
Changed to 
```
Time [us]: 449 (instantiation: 196, execution: 253)
```

**Comment by axic** [comment]
Tests seem to fail?

**Comment by chfast** [comment]
All green now.

**Comment by axic** [comment]
I still don't like the perhaps non-existent overhead, but don't have the time to "improve" it.

**Pull request merged**